First-cut implementation of the codegen

hideya opened the pull request:
Based on the discussions on PR https://github.com/strongloop/loopback-sdk-ios/pull/59, the codegen is repackaged into this new repository.

@raymondfeng , @bajtos does this look good to you folks?

hideya committed:
First-cut implementation of the codegen.

hideya commented:
Thanks, @raymondfeng , for the review!
I've fixed three issues pointed out below and cleaned up the commit history.
https://github.com/hideya/loopback-sdk-ios-codegen/commit/3d1e1eaf629ca7d2090874fc1fc5770d6b303a67#commitcomment-13564713

bajtos commented on README.md:
```diff
@@ -1,2 +1,34 @@
+The following is an example of the usage applied to the test server provided under this directory:
+
+ 1. Run `npm install` in this directory to initialize the tool `bin/codegen-ios`.
+ *  Run `pushd test-env; npm install; popd` to initialize the test server.
```
I prefer to keep test dependencies in project's master `package.json` as `devDependencies`. That way a single `npm install` is enough.

hideya replied:
Thanks for pointing it out!  I've updated the master `package.json` and removed the child.

bajtos commented on bin/codegen-ios:
```diff
@@ -0,0 +1,84 @@
+function assertLoopBackVersion(app) {
+  // app.loopback property was added before v2.0.0.
+  // If app.loopback is undefined, then you can safely assume it's an unsupported version.
+  if (typeof app.loopback === 'undefined') {
```
Note that `app.loopback` was defined before `v2.0`, this check will pass for `v1.9.x`. I am proposing to use [semver](https://www.npmjs.com/package/semver) to check whether loopback versions matches our supported range.

hideya replied:
Thanks for the careful review.  I'll update the code following your suggestion.

bajtos commented on package.json:
```diff
@@ -0,0 +1,28 @@
+  "description": "Tool for auto-generating iOS client source code for LoopBack App",
+  "keywords": [ "StrongLoop Labs", "LoopBack", "client", "SDK", "iOS", "codegen" ],
+  "bin": {
+    "codegen-ios": "bin/codegen-ios"
```
The Angular SDK uses `lb-ng` command, please use the `lb-` prefix here too. I am personally fine with the short name `lb-ios`, although `lb-codegen-ios` is ok too.

hideya replied:
Sure, I'll update it to `lb-ios`.

bajtos commented on test-env/server/boot/explorer.js:
```diff
@@ -0,0 +1,27 @@
+    var explorerPath = explorerApp.mountpath || explorerApp.route;
+    console.log('Browse your REST API at %s%s', baseUrl, explorerPath);
+  });
+};
```
We use slightly different project layout now and both explorer & rest api are loaded via JSON files (see https://github.com/strongloop/loopback-workspace/pull/230). Could you please update the configuration of the test project to follow the current convention?

hideya replied:
Sure, I will!  Thanks for the pointer to the discussion.

bajtos commented on test-env/package.json:
```diff
@@ -0,0 +1,27 @@
+    "serve-favicon": "^2.0.1"
+  },
+  "optionalDependencies": {
+    "loopback-explorer": "^1.1.0"
```
Note: when you upgrade to the current project structure, this will become a regular dependency on `loopback-component-explorer` version `^2.1.1`

hideya replied:
Thanks for the heads-up!  I'll make it sure.

hideya commented:
Thanks, @bajtos , for the review!
I've addressed all the points (the following items).  Looks good?

- https://github.com/strongloop/loopback-sdk-ios-codegen/pull/1/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R29
- https://github.com/strongloop/loopback-sdk-ios-codegen/pull/1/files#diff-d037562e434d4638d5a12f2ecf491154R79
- https://github.com/strongloop/loopback-sdk-ios-codegen/pull/1/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R7
- https://github.com/strongloop/loopback-sdk-ios-codegen/pull/1/files#diff-06bc9e479f96ee126e1313751d7902f0R2
- https://github.com/strongloop/loopback-sdk-ios-codegen/pull/1/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R34

bajtos commented on README.md:
```diff
@@ -1,2 +1,33 @@
+After performing the above steps, run the followings:
+
+ 1. Run the test server by `node test-env/server/server.js &`.
+ *  Copy `LoopBack.framework` into `test-env/CodeGenTest/`.
```
Is there a way how to reference source files from LoopBack.framework directly, so that it is not necessary to copy them?

If not, then I am proposing the create an npm script to run the generator and copy the files in one step.

Intended usage: `npm run test-setup`

Implementation - add a new script entry to package.json:

``` json
{
  "scripts": {
    "test-setup": "bin/lb-ios -p XX test-env/server/server test-env/client/io/gen-src && cp -R LoopBack.framework test-env/CodeGenTest"
  }
}
```

Thoughts?

hideya replied:
Could I deal with this a bit later once the entire structure becomes stabler?

I guess my explanation on this codegen was a bit too short and unkind.
`LoopBack.framework` is the one generated from `loopback-sdk-ios` (it is a package of the lib file and the headers).
The usage of this codegen I intended was that, instead of dealing with all the SDK in the shape of source files like Xamarin SDK, developers incorporates precompiled core part (including SLRemoting, SLAFNetworking and basic LBModel classes) as an iOS framework and compile their app with this framework together with the generated code that is specific to their app.  I thought this is simpler for the developers. 
Does this make sense?
If this is the way to go, maybe I should use CocoaPods for the dependency management.

bajtos replied:
> Could I deal with this a bit later once the entire structure becomes stabler?

Sure, whatever works best for you.

> LoopBack.framework is the one generated from loopback-sdk-ios (it is a package of the lib file and the headers).
> The usage of this codegen I intended was that, instead of dealing with all the SDK in the shape of source files like Xamarin SDK, developers incorporates precompiled core part (including SLRemoting, SLAFNetworking and basic LBModel classes) as an iOS framework and compile their app with this framework together with the generated code that is specific to their app. I thought this is simpler for the developers. 
> Does this make sense?
> If this is the way to go, maybe I should use CocoaPods for the dependency management.

Oh, I did not realise `LoopBack.framework` comes from a different repository. In which case yes, it would be best to switch to CocoaPods in the long term.

As a quick short-term fix, could you please simply extend this line and add instructions on where to get that `LoopBack.framework` directory from? Are users supposed to download an archive from https://docs.strongloop.com/display/LB/iOS+SDK? Clone the directory from https://github.com/strongloop/loopback-sdk-ios? Something else?

hideya replied:
Sure, I'll update the usage guide to include where to get the framework.
Since I needed some updates in the SDK to make the codegen work fully, the framework needs to come from  https://github.com/strongloop/loopback-sdk-ios.  I'll add this to the README.

bajtos replied:
Cool!

> Since I needed some updates in the SDK to make the codegen work fully, the framework needs to come from https://github.com/strongloop/loopback-sdk-ios.

We can (and should!) make a new release of the SDK. I think @raymondfeng may know the release process.

hideya replied:
Sounds good!  I guess I'll ping @raymondfeng when I finish the first cut of method codegen as I'm finding (minor) improvements in the iOS SDK code as I work on the method support (I'm accumulating the fixes in this PR: https://github.com/strongloop/loopback-sdk-ios/pull/67)  I may also work on Buffer and GeoPoint types support before asking the release (I've added Date type support but not yet worked on those two).

bajtos commented:
LGTM, let's get this landed. You can address my comment above later (or not at all).
